Q-municate/Classes/Core/QMPushNotificationManager/QMPushNotificationManager.m

```diff
@@ -12,51 +12,193 @@
 
 static NSString * const kQMNotificationActionTextAction = @"TEXT_ACTION";
 static NSString * const kQMNotificationCategoryReply = @"TEXT_REPLY";
+static NSString * const kQMSubscriptionKey = @"SUBSCRIPTION_KEY";
+static NSString * const kQMDeviceTokenKey = @"DEVICE_TOKEN_KEY";
 
+typedef void(^QBTokenCompletionBlock)(NSData *token, NSError *error);
 
-@interface QMPushNotificationManager ()
+@interface QMPushNotificationManager () {
+    QBMSubscription *internalSubscription;
+    NSData *internalToken;
+}
 
 @property (weak, nonatomic) QMCore <QMServiceManagerProtocol>*serviceManager;
+@property (copy, nonatomic) QBTokenCompletionBlock tokenCompletionBlock;
+
 
 @end
 
 @implementation QMPushNotificationManager
 
 @dynamic serviceManager;
 
+
+- (BFTask *)unregisterFromPushNotificationsAndUnsubscribe:(BOOL)shouldUnsubscribe {
+    
```

Comment:
should be:

 [[UIApplication sharedApplication] unregisterForRemoteNotifications];

[[self unSubscribeFromPushNotifications] continueWithBlock:^id _Nullable(BFTask * _Nonnull t {
             if (t.isCancelled) {
                 [source trySetCancelled];
             }
             else if (t.error) {
                 [source setError:t.error];
         }];